botan3: fix static build

WilliButz opened the pull request:
Fixes static build of the library, which broke with the move to pkgs/by-name in #446190 (98a16089ef58d31311fdc177fdf07189658b3fb8).
Also fixes the static cli build using the same workaround for the gcc_eh linking issue as in #391424, which did not produce a statically linked binary even before the move to pkgs/by-name.

(Note that both builds previously ran through without errors, but neither was usable as intended, the library archive wasn't usable for static linking and the cli was dynamically linked)

Also see commit messages.

## Things done

- Built on platform:
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [x] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

---

WilliButz committed:
botan3: fix static cli build through empty-libgcc_eh workaround

Prior to this commit the build of `pkgsStatic.botan3.bin` produced a
dynamically linked binary, indenpendently of the move to
`pkgs/by-name`. Previous commit message and comments also suggest this
it was expected for the static build to only produce the static library,
not the statically linked cli.

wolfgangwalther reviewed (dismissed):
Clear no on moving back out of by-name without justification - I think it is not actually needed in this case.

Please note that @emilazy and I are working hard on moving almost *everything* into by-name in the next cycle. Thus, if there is a reason that by-name currently doesn't work here, we're happy to discuss it and see whether we can fix it here or not, yet.

WilliButz commented on pkgs/top-level/all-packages.nix:
```diff
@@ -6976,6 +6975,10 @@ with pkgs;
+  botan3 = callPackage ../development/libraries/botan {
+    # botan3 only sensibly works with libcxxStdenv when building static binaries
+    stdenv = if stdenv.hostPlatform.isStatic then libcxxStdenv else stdenv;
+  };
```
Yes this is the critical part. So you're saying to keep it in by-name *and* add it to top-level with callPackage on the package.nix?

WilliButz committed:
botan3: fix selection of stdenv variant for static builds

Apparently with the move to by-name broke stdenv selection in such a way
that the resulting library archive from `pkgsStatic.botan3` no longer
was suitable for static linking, resulting in linking errors:

```
/nix/store/<...>-x86_64-unknown-linux-musl-binutils-2.44/bin/x86_64-unknown-linux-musl-ld: cannot find -lbz2: No such file
/nix/store/<...>-x86_64-unknown-linux-musl-binutils-2.44/bin/x86_64-unknown-linux-musl-ld: have you installed the static ve
/nix/store/<...>-x86_64-unknown-linux-musl-binutils-2.44/bin/x86_64-unknown-linux-musl-ld: cannot find -ljitterentropy: No
/nix/store/<...>-x86_64-unknown-linux-musl-binutils-2.44/bin/x86_64-unknown-linux-musl-ld: have you installed the static ve
/nix/store/<...>-x86_64-unknown-linux-musl-binutils-2.44/bin/x86_64-unknown-linux-musl-ld: cannot find -lz: No such file or
/nix/store/<...>-x86_64-unknown-linux-musl-binutils-2.44/bin/x86_64-unknown-linux-musl-ld: have you installed the static ve
```

Co-authored-by: Maximilian Bosch <maximilian@mbosch.me>

nikstur reviewed (approved):
c8a230069b95520abc24a2c3daa51956e21411a8 is wild. Should really be fixed in a more general way. Cool that you found this workaround though!

Ma27 commented:
@nikstur when we debugged the problem on Friday, the underlying issue strongly looked like an LLVM bug to me: https://github.com/llvm/llvm-project/issues/32805

I.e. when using GNU tooling for linking (which is the case here) then clang generates the wrong set of flags (wrong as in gcc_eh doesn't exist for a GCC that's built with `--disable-shared`).
